ValveResourceFormat/IO/ModelExtract.cs

```diff
@@ -0,0 +1,613 @@
+                )
+            ));
+        }
+
+        void RemapMaterials(
+            IReadOnlyDictionary<string, string> remapTable = null,
+            bool globalReplace = false,
+            string globalDefault = "materials/tools/toolsnodraw.vmat")
+        {
+            var remaps = new KVObject(null, isArray: true);
+            materialGroupList.Value.AddProperty(null, new KVValue(KVType.OBJECT,
+                MakeNode(
+                    "DefaultMaterialGroup",
+                    ("remaps", new KVValue(KVType.ARRAY, remaps)),
+                    ("use_global_default", new KVValue(KVType.BOOLEAN, globalReplace)),
+                    ("global_default_material", new KVValue(KVType.STRING, globalDefault))
+                )
+            ));
+
+            if (globalReplace || remapTable == null)
+            {
+                return;
+            }
+
+            foreach (var (from, to) in remapTable)
+            {
+                var remap = new KVObject(null);
+                remap.AddProperty("from", new KVValue(KVType.STRING, from));
+                remap.AddProperty("to", new KVValue(KVType.STRING, to));
+                remaps.AddProperty(null, new KVValue(KVType.OBJECT, remap));
+            }
+        }
+    }
+
+    public string GetFileName()
+        => model?.Data.GetProperty<string>("m_name")
+            ?? fileName;
+
+    string GetDmxFileName_ForEmbeddedMesh(string subString, int number = 0)
+    {
+        var fileName = GetFileName();
+        return (Path.GetDirectoryName(fileName)
+            + Path.DirectorySeparatorChar
+            + Path.GetFileNameWithoutExtension(fileName)
+            + "_"
+            + subString
+            + (number > 0 ? number : string.Empty)
+            + ".dmx")
+            .Replace('\\', '/');
+    }
+
+    static string GetDmxFileName_ForReferenceMesh(string fileName)
+        => Path.ChangeExtension(fileName, ".dmx").Replace('\\', '/');
+
+    private void EnqueueRenderMeshes()
+    {
+        if (model == null)
+        {
+            return;
+        }
+
+        var i = 0;
+        foreach (var embedded in model.GetEmbeddedMeshes())
+        {
+            RenderMeshesToExtract.Add((embedded.Mesh, GetDmxFileName_ForEmbeddedMesh(embedded.Name, i++)));
+        }
+
+        foreach (var reference in model.GetReferenceMeshNamesAndLoD())
+        {
+            using var resource = fileLoader.LoadFile(reference.MeshName + "_c");
+            RenderMeshesToExtract.Add(((Mesh)resource.DataBlock, GetDmxFileName_ForReferenceMesh(reference.MeshName)));
+        }
+    }
+
+    private void EnqueuePhysMeshes()
+    {
+        if (physAggregateData == null)
+        {
+            return;
+        }
+
+        PhysicsSurfaceNames = physAggregateData.SurfacePropertyHashes.Select(hash =>
+        {
+            StringToken.InvertedTable.TryGetValue(hash, out var name);
+            return name ?? hash.ToString(CultureInfo.InvariantCulture);
+        }).ToArray();
```

Comment:
Populate StringToken with names from surfaceproperties.vsurf for better hit rate. Alternatively the world_physics.vmdl RED2 has the used properties listed as well.